src/Extension/PublishableExtension.php

```diff
@@ -0,0 +1,131 @@
+                    $queryBuilder->expr()->isNotNull("$alias.$configuration->fieldName"),
+                    $queryBuilder->expr()->lte("$alias.$configuration->fieldName", ':currentTime'),
+                ),
+                $queryBuilder->expr()->andX(
+                    $queryBuilder->expr()->orX(
+                        $queryBuilder->expr()->isNull("$alias.$configuration->fieldName"),
+                        $queryBuilder->expr()->gt("$alias.$configuration->fieldName", ':currentTime'),
+                    ),
+                    $queryBuilder->expr()->eq("$alias.$configuration->associationName", ":id_$identifier"),
+                )
+            )->setParameter('currentTime', new \DateTimeImmutable())
+            ->setParameter("id_$identifier", $identifier);
+        }
+    }
+
+    public function applyToCollection(QueryBuilder $queryBuilder, QueryNameGeneratorInterface $queryNameGenerator, string $resourceClass, string $operationName = null, array $context = []): void
+    {
+        if (!$configuration = $this->getConfiguration($resourceClass)) {
+            return;
+        }
+
+        $configuration = $this->getConfiguration($resourceClass);
+        if (!$this->isAllowed($context)) {
+            // User has no access to draft object
+            $this->updateQueryBuilderForUnauthorizedUsers($queryBuilder, $configuration);
+
+            return;
+        }
+
+        $alias = $queryBuilder->getRootAliases()[0];
+        $publishedResourceAlias = $queryNameGenerator->generateJoinAlias($configuration->associationName);
+        $queryBuilder->leftJoin("$alias.$configuration->associationName", $publishedResourceAlias);
+
+        // (o.publishedAt IS NOT NULL AND o.publishedAt <= :currentTime) OR (o.publishedAt IS NULL OR o.publishedAt > :currentTime)
+        $queryBuilder->orWhere(
+            $queryBuilder->expr()->andX(
+                $queryBuilder->expr()->isNotNull("$alias.$configuration->fieldName"),
+                $queryBuilder->expr()->lte("$alias.$configuration->fieldName", ':currentTime'),
+            ),
+            $queryBuilder->expr()->orX(
+                $queryBuilder->expr()->isNull("$alias.$configuration->fieldName"),
+                $queryBuilder->expr()->gt("$alias.$configuration->fieldName", ':currentTime'),
+            ),
+        )->setParameter('currentTime', new \DateTimeImmutable());
+    }
+
+    private function isAllowed(array $context): bool
+    {
+        return $this->publishableHelper->isGranted() && false === ($context['filters']['published'] ?? false);
+    }
+
+    private function updateQueryBuilderForUnauthorizedUsers(QueryBuilder $queryBuilder, Publishable $configuration): void
+    {
+        $alias = $queryBuilder->getRootAliases()[0];
+        $queryBuilder
+            ->andWhere("$alias.$configuration->fieldName IS NOT NULL")
+            ->andWhere("$alias.$configuration->fieldName >= :currentTime")
```

Comment:
should be `<=` (less than) I think... Database value of publishedAt should be in the past to restrict to published only resources